algebird-core/src/main/scala/com/twitter/algebird/CountMinSketch.scala

```diff
@@ -441,23 +441,90 @@ case class CMSZero[K](override val params: CMSParams[K]) extends CMS[K](params)
  */
 case class CMSItem[K](item: K, override val totalCount: Long, override val params: CMSParams[K]) extends CMS[K](params) {
 
-  override def +(x: K, count: Long): CMS[K] = CMSInstance[K](params) + item + (x, count)
+  override def +(x: K, count: Long): CMS[K] = SparseCMS[K](params) + (item, totalCount) + (x, count)
 
   override def ++(other: CMS[K]): CMS[K] = {
     other match {
       case other: CMSZero[_] => this
-      case other: CMSItem[K] => CMSInstance[K](params) + item + other.item
-      case other: CMSInstance[K] => other + item
+      case other: CMSItem[K] => CMSInstance[K](params) + (item, totalCount) + (other.item, other.totalCount)
+      case _ => other + item
     }
   }
 
-  override def frequency(x: K): Approximate[Long] = if (item == x) Approximate.exact(1L) else Approximate.exact(0L)
+  override def frequency(x: K): Approximate[Long] = if (item == x) Approximate.exact(totalCount) else Approximate.exact(0L)
 
-  override def innerProduct(other: CMS[K]): Approximate[Long] = other.frequency(item)
+  override def innerProduct(other: CMS[K]): Approximate[Long] = Approximate.exact(totalCount) * other.frequency(item)
 
 }
 
 /**
+ * A sparse Count-Min sketch structure, used for situations where the key is highly skewed.
+ */
+case class SparseCMS[K](exactCountTable: Map[K, Long], maxExactCountOpt: Option[Int] = None,
```

Comment:
let's put the maxExactCountOpt into CMSParams. Without that, you can't really control it, right? Since CMSItem will always create with the default setting, right?